### PR TITLE
Precalculate contract address function

### PR DESCRIPTION
**Pull request by tkumor3**
Closes #

## Introduced changes

-
-

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: benchmarks/data/forge.cairo [hunk 1]
```diff
@@ -9,12 +9,13 @@ use cheatcodes::PreparedContract;
...
+
 fn deploy_hello_starknet() -> ContractAddress {
-    let class_hash = declare('HelloStarknet').unwrap();
-    let prepared = PreparedContract {
-        contract_address: 1234, class_hash: class_hash, constructor_calldata: @ArrayTrait::new()
-    };
-    let contract_address = deploy(prepared).unwrap();
+    let contract = declare('HelloStarknet').unwrap();
+    let constructor_calldata = @ArrayTrait::new()
+    let contract_address = contract.deploy(prepared, constructor_calldata).unwrap();
```
Comment (piotmag769) [inline benchmarks/data/forge.cairo:18]:
```suggestion
    let contract_address = contract.deploy(constructor_calldata).unwrap();
```

File: cheatcodes/src/cheatcodes.cairo [hunk 2]
```diff
@@ -34,54 +44,81 @@ impl RevertedTransactionImpl of RevertedTransactionTrait {
...
+        };
 
-fn deploy(prepared_contract: PreparedContract) -> Result::<ContractAddress, RevertedTransaction> {
-    let PreparedContract{class_hash, constructor_calldata } = prepared_contract;
-    let mut inputs = array![class_hash.into()];
+        let outputs = cheatcode::<'precalculate_address'>(inputs.span());
+        let result = *outputs[0];
+        result.try_into().unwrap()
+    }
+    fn deploy(self: @ContractClass, constructor_calldata: @Array::<felt252>) -> Result<ContractAddress, RevertedTransaction> {
```
Comment (piotmag769) [inline cheatcodes/src/cheatcodes.cairo:69]: newline

**Review by piotmag769 — COMMENTED** [review]
We can remove `PrepraredContract` now too

File: crates/cheatable-starknet/src/lib.rs [hunk 3]
```diff
@@ -20,8 +23,17 @@ impl CheatedState {
...
+            deploy_counter: 0,
         }
     }
+
+    pub fn increment_deploy_counter(&mut self) -> () {
+        self.deploy_counter += 1;
+    }
+
+    pub fn gen_salt(&self) -> ContractAddressSalt {
+        ContractAddressSalt(StarkFelt::from(333_u32 + self.deploy_counter))
```
Comment (tkumor3) [inline crates/cheatable-starknet/src/lib.rs:35]: Let me know is it ok? @piotmag769 @Arcticae

File: crates/forge/tests/data/simple_package/tests/ext_function_test.cairo [hunk 4]
```diff
@@ -1,12 +1,20 @@
...
-}
+    let contract = declare('HelloStarknet');
+    let contract_address_pre = contract.precalculate_address(@calldata);
+    let contract_address = contract.deploy(@calldata).unwrap();
+    let contract_address_pre2 = contract.precalculate_address(@calldata);
+    let contract_address2 = contract.deploy(@calldata).unwrap();
+
+    assert(contract_address_pre.into() == contract_address, contract_address.into());
+    assert(contract_address_pre2.into() == contract_address2, contract_address.into());
+}
```
Comment (piotmag769) [inline crates/forge/tests/data/simple_package/tests/ext_function_test.cairo:20]: newline

File: crates/forge/tests/data/simple_package/tests/ext_function_test.cairo
Comment (piotmag769) [inline crates/forge/tests/data/simple_package/tests/ext_function_test.cairo:1]: I don't think we want to change this particular e2e test, integration tests are enough

File: crates/cheatable-starknet/src/cheatcodes/deploy.rs [hunk 5]
```diff
@@ -42,14 +42,16 @@ impl CheatedState {
         for felt in inputs.iter().skip(2).take(calldata_length) {
             calldata.push(felt.clone());
         }
-
+        // self.increment_deploy_counter();
+        // print!("deploy_counter {}", self.deploy_counter);
```
Comment (piotmag769) [inline crates/cheatable-starknet/src/cheatcodes/deploy.rs:46]: commented out code

File: cheatcodes/src/cheatcodes.cairo [hunk 6]
```diff
@@ -34,54 +44,80 @@ impl RevertedTransactionImpl of RevertedTransactionTrait {
...
+        let calldata_len = constructor_calldata.len();
+        let mut i = 0;
+
+        loop {
+            if i == calldata_len {
+                break();
+            }
+            inputs.append(*constructor_calldata[i]);
+            i += 1;
+        };
```
Comment (war-in) [inline cheatcodes/src/cheatcodes.cairo:63]: I think this code is duplicated in `deploy` method

File: crates/cheatable-starknet/src/cheatcodes/deploy.rs (hunk 5 above)
Comment (war-in) [inline crates/cheatable-starknet/src/cheatcodes/deploy.rs:46]: 👀

File: crates/cheatable-starknet/src/cheatcodes/deploy.rs [hunk 7]
```diff
@@ -42,14 +42,16 @@ impl CheatedState {
...
         // Deploy a contract using syscall deploy.
-        let account_address = ContractAddress(patricia_key!(TEST_ACCOUNT_CONTRACT_ADDRESS));
+        let account_address: ContractAddress =
```
Comment (war-in) [inline crates/cheatable-starknet/src/cheatcodes/deploy.rs:48]: Are you sure we have to specify the type?

File: crates/cheatable-starknet/src/cheatcodes/deploy.rs [hunk 8]
```diff
@@ -42,14 +42,16 @@ impl CheatedState {
...
+            ContractAddress(patricia_key!(TEST_ACCOUNT_CONTRACT_ADDRESS));
         let block_context = build_block_context();
         let entry_point_selector = selector_from_name("deploy_contract");
-        let salt = ContractAddressSalt::default();
+        let salt = self.gen_salt();
         let class_hash = ClassHash(StarkFelt::new(class_hash.to_be_bytes()).unwrap());
-
+        self.increment_deploy_counter();
```
Comment (piotmag769) [inline crates/cheatable-starknet/src/cheatcodes/deploy.rs:54]: I wonder if we should increment it here or only after successful deploys
cc @Arcticae

File: crates/cheatable-starknet/src/lib.rs (hunk 3 above)
Comment (piotmag769) [reply crates/cheatable-starknet/src/lib.rs:35]: Seems fine

File: crates/cheatable-starknet/src/cheatcodes/precalculate_address.rs [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+impl CheatedState {
+    pub fn precalculate_address(
+        &self,
+        buffer: &mut MemBuffer,
+        inputs: &[Felt252],
+    ) -> Result<(), EnhancedHintError> {
+        let class_hash = inputs[0].clone();
+
+        let account_address: ContractAddress =
```
Comment (war-in) [inline crates/cheatable-starknet/src/cheatcodes/precalculate_address.rs:22]: same as above

**Review by piotmag769 — COMMENTED** [review]
Good job! Please update `CHANGELOD.md` and make sure to rebase to master branch after #419 is merged (blockifier operations and writing to memory got separated)

File: crates/cheatable-starknet/src/cheatcodes/precalculate_address.rs [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+            ContractAddress(patricia_key!(TEST_ACCOUNT_CONTRACT_ADDRESS));
+        let class_hash = ClassHash(StarkFelt::new(class_hash.to_be_bytes()).unwrap());
+        let salt = self.gen_salt();
+        let calldata_length = inputs[1].to_usize().unwrap();
+
+        let mut calldata = vec![];
+        for felt in inputs.iter().skip(2).take(calldata_length) {
+            calldata.push(felt.clone());
+        }
```
Comment (war-in) [inline crates/cheatable-starknet/src/cheatcodes/precalculate_address.rs:31]: I believe we can do something like 
```
let mut calldata = Vec::from(&inputs[2..(2 + calldata_length)]);
```

File: crates/cheatable-starknet/src/cheatcodes/precalculate_address.rs [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+            .expect("Failed to insert declared contract class hash");
+        Ok(())
+    }
+}
+
+fn create_execute_calldata(calldata: &[Felt252]) -> Calldata {
+    let mut execute_calldata = vec![];
+    let mut calldata: Vec<StarkFelt> = calldata
+        .iter()
+        .map(|data| StarkFelt::new(data.to_be_bytes()).unwrap())
```
Comment (war-in) [inline crates/cheatable-starknet/src/cheatcodes/precalculate_address.rs:55]: afaik we have `felt_to_stark_felt` method

File: crates/cheatable-starknet/src/lib.rs (hunk 3 above)
Comment (war-in) [reply crates/cheatable-starknet/src/lib.rs:35]: Why do we use `333`? :hmmm:

File: crates/cheatable-starknet/src/cheatcodes/precalculate_address.rs [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+        let execute_calldata = create_execute_calldata(&calldata);
+        let contract_address =
+            calculate_contract_address(salt, class_hash, &execute_calldata, account_address)
+                .unwrap();
+
+        let contract_address = Felt252::from_bytes_be((*contract_address.0.key()).bytes());
+
+        buffer
+            .write(Felt252::from(0))
+            .expect("Failed to insert error code");
```
Comment (piotmag769) [inline crates/cheatable-starknet/src/cheatcodes/precalculate_address.rs:42]: I think we should not write the error code here since this function should always succeed, it only makes sense when returning `Result` in Cairo. I think we do that in `declare.rs` as well though, could you change that too? Remember to change corresponding Cairo code

File: crates/forge/tests/integration/precalculate_address.rs [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+        use traits::TryInto;
+        use starknet::ContractAddressIntoFelt252;
+
+        #[starknet::interface]
+        trait IHelloStarknet<TContractState> {
+            fn increase_balance(ref self: TContractState, amount: felt252);
+            fn get_balance(self: @TContractState) -> felt252;
+            fn do_a_panic(self: @TContractState);
+            fn do_a_panic_with(self: @TContractState, panic_data: Array<felt252>);
+        }
```
Comment (piotmag769) [inline crates/forge/tests/integration/precalculate_address.rs:27]: You don't need to copy the interface here since we do not use it in this test anyways

File: crates/forge/src/cheatcodes_hint_processor.rs [hunk 14]
```diff
@@ -236,6 +236,9 @@ impl CairoHintProcessor<'_> {
                 print(inputs);
                 Ok(())
             }
+            "precalculate_address" => self
+                .cheated_state
+                .precalculate_address(&mut buffer, &inputs),
```
Comment (war-in) [inline crates/forge/src/cheatcodes_hint_processor.rs:241]: The newest standard (from the https://github.com/foundry-rs/starknet-foundry/pull/419 which is going to be merged in 30 min) extracts inputs here and passes them as a separate values

File: crates/forge/tests/data/simple_package/tests/ext_function_test.cairo
Comment (tkumor3) [reply crates/forge/tests/data/simple_package/tests/ext_function_test.cairo:1]: I added it by mistake

File: crates/forge/tests/integration/precalculate_address.rs
Comment (piotmag769) [inline crates/forge/tests/integration/precalculate_address.rs:1]: I think we should also test it with contract with non-empty calldata so we can check what happens when different calldata is being used. For example (the code is ugly, but the idea is there xD):

```cairo
let contract = declare('X');
let c1 = array![1, 2];
let c2 = array![3, 4];
let addr1_pre: felt252 = contract.precalculate_address(@c1).into();
let addr2_pre: felt252 = contract.precalculate_address(@c2).into();
let addr1: felt252 = contract.deploy(@c1).into();
let addr2_after: felt252 = contract.precalculate_address(@c2).into();
let addr2: felt252 = contract.deploy(@c2).into();
assert(addr1_pre != addr_2_pre && addr1 == addr1_pre && addr2 == addr2_after && addr2 != addr2_pre, '')
```
or sth like that

File: crates/forge/tests/integration/precalculate_address.rs [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+            let mut calldata = ArrayTrait::new();
+
+            let contract = declare('HelloStarknet');
+            let contract_address_pre = contract.precalculate_address(@calldata);
+            let contract_address = contract.deploy(@calldata).unwrap();
+            let contract_address_pre2 = contract.precalculate_address(@calldata);
+            let contract_address2 = contract.deploy(@calldata).unwrap();
+
+            assert(contract_address_pre.into() == contract_address, contract_address.into());
+            assert(contract_address_pre2.into() == contract_address2, contract_address.into());
```
Comment (piotmag769) [inline crates/forge/tests/integration/precalculate_address.rs:40]: Assert that `contract_address != contract_address2`

**Review by piotmag769 — COMMENTED** [review]
Remember to get rid of `PreparedContract`

File: crates/cheatable-starknet/src/cheatcodes/deploy.rs (hunk 8 above)
Comment (Arcticae) [reply crates/cheatable-starknet/src/cheatcodes/deploy.rs:54]: I think only after successful ones, but it doesn't matter that much (i guess?)

File: crates/cheatable-starknet/src/cheatcodes/precalculate_address.rs (hunk 10 above)
Comment (piotmag769) [reply crates/cheatable-starknet/src/cheatcodes/precalculate_address.rs:31]: I believe my code was an inspiration for that and it is in "deploy" match arm too xD If that's true we could change it there as well

File: crates/forge/tests/integration/precalculate_address.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (piotmag769) [reply crates/forge/tests/integration/precalculate_address.rs:40]: Hmm, maybe PartialEq is implemented for `ContractAddress`, didn't know that, thx!

File: benchmarks/data/forge.cairo [hunk 16]
```diff
@@ -9,12 +9,13 @@ use cheatcodes::PreparedContract;
 use project::IHelloStarknetSafeDispatcher;
 use project::IHelloStarknetSafeDispatcherTrait;
 
+use cheatcodes::ContractClass;
+use cheatcodes::ContractClassTrait;
```
Comment (Arcticae) [inline benchmarks/data/forge.cairo:12]: I think you don't have to import those, right?

File: cheatcodes/src/cheatcodes.cairo [hunk 17]
```diff
@@ -34,54 +44,74 @@ impl RevertedTransactionImpl of RevertedTransactionTrait {
...
+    let span = cheatcode::<'declare'>(array![contract].span());
+
+    let exit_code = *span[0];
+    let result = *span[1];
+    assert(exit_code == 0, 'declare should never fail');
+    let class_hash = result.try_into().unwrap();
+
+    ContractClass {
+        class_hash: class_hash
     }
```
Comment (Arcticae) [inline cheatcodes/src/cheatcodes.cairo:109]: `ContractClass {
        class_hash
}`
is sufficient

File: cheatcodes/src/cheatcodes.cairo [hunk 18]
```diff
@@ -34,54 +44,74 @@ impl RevertedTransactionImpl of RevertedTransactionTrait {
...
-}
+        let outputs = cheatcode::<'precalculate_address'>(inputs.span());
+        (*outputs[1]).try_into().unwrap()
+    }
 
-fn deploy(prepared_contract: PreparedContract) -> Result::<ContractAddress, RevertedTransaction> {
-    let PreparedContract{class_hash, constructor_calldata } = prepared_contract;
-    let mut inputs = array![class_hash.into()];
+    fn deploy(self: @ContractClass, constructor_calldata: @Array::<felt252>) -> Result<ContractAddress, RevertedTransaction> {
+        let class_hash: felt252 = (*self.class_hash).into();
```
Comment (Arcticae) [inline cheatcodes/src/cheatcodes.cairo:51]: I think you can perform conversion inside of `_prepare_calldata` function instead

File: cheatcodes/src/lib.cairo [hunk 19]
```diff
@@ -1,11 +1,13 @@
...
 
 use cheatcodes::declare;
-use cheatcodes::deploy;
 
 use cheatcodes::PreparedContract;
 use cheatcodes::RevertedTransaction;
 use cheatcodes::RevertedTransactionTrait;
+use cheatcodes::ContractClass;
+use cheatcodes::ContractClassTrait;
+use cheatcodes::ContractClassImpl;
```
Comment (Arcticae) [inline cheatcodes/src/lib.cairo:9]: I don't think you need to import the impl? I might be wrong though.

File: crates/cheatable-starknet/src/lib.rs [hunk 20]
```diff
@@ -20,8 +23,17 @@ impl CheatedState {
...
             warped_contracts: HashMap::new(),
+            deploy_counter: 0,
         }
     }
+
+    pub fn increment_deploy_counter(&mut self) -> () {
+        self.deploy_counter += 1;
+    }
+
+    pub fn gen_salt(&self) -> ContractAddressSalt {
```
Comment (Arcticae) [inline crates/cheatable-starknet/src/lib.rs:34]: I think this could be `generate_salt` (for sake of being explicit), and it could also automatically increment so we don't forget to do it manually.

File: crates/cheatable-starknet/src/lib.rs (hunk 3 above)
Comment (Arcticae) [reply crates/cheatable-starknet/src/lib.rs:35]: Can't this be just `deploy_counter` ?

File: crates/cheatable-starknet/src/lib.rs [hunk 21]
```diff
@@ -20,8 +23,17 @@ impl CheatedState {
             rolled_contracts: HashMap::new(),
             pranked_contracts: HashMap::new(),
             warped_contracts: HashMap::new(),
+            deploy_counter: 0,
```
Comment (Arcticae) [inline crates/cheatable-starknet/src/lib.rs:26]: I would rename to `salt_base` or `deploy_salt_base` or something like that

**Review by Arcticae — APPROVED** [review]
A few nitpicks - but generally LGTM, good job!

File: docs/src/appendix/forge-library/declare.md [hunk 22]
```diff
@@ -1,8 +1,8 @@
 # `declare`
 
-> `fn declare(contract: felt252) -> ClassHash`
+> `fn declare(contract: felt252) -> ContractClass`
 
-Declares a contract and returns its class hash.
+Declares a contract and returns `ContractClass`.
```
Comment (piotmag769) [inline docs/src/appendix/forge-library/declare.md:5]: Would be nice to describe what `ContractClass` is here as well - a struct with xyz fields and that you can call precalculate_address and deploy on its instance.

File: docs/src/appendix/forge-library/deploy.md [hunk 23]
```diff
@@ -1,31 +1,34 @@
 # `deploy`
 
-> `fn deploy(prepared_contract: PreparedContract) -> Result::<ContractAddress, RevertedTransaction>`
+```rust 
+trait ContractClassTrait {
+    fn deploy(self: @ContractClass, constructor_calldata: @Array::<felt252>) -> Result<ContractAddress, RevertedTransaction>;
+    ...
```
Comment (piotmag769) [inline docs/src/appendix/forge-library/deploy.md:6]:
```suggestion
    // ...
```

File: docs/src/appendix/forge-library/deploy.md [hunk 24]
```diff
@@ -1,31 +1,34 @@
...
+}
+```
 
 Deploys a contract and returns its address.
 
-- `prepared_contract` - an instance of the struct `PreparedContract` that consists of the following fields:
+- `self` - an instance of the struct `ContractClass` that consists of the following fields:
```
Comment (piotmag769) [inline docs/src/appendix/forge-library/deploy.md:12]: We could mention that it is obtained by calling `declare`

File: docs/src/testing/contracts.md [hunk 25]
```diff
@@ -118,7 +117,7 @@ If we called this function in a test, it would result in a failure.
 fn failing() {
     // ...
 
-    let contract_address = deploy(prepared).unwrap();
+    let contract_address = contract.deploy(calldata).unwrap();
```
Comment (piotmag769) [inline docs/src/testing/contracts.md:120]:
```suggestion
    let contract_address = contract.deploy(@calldata).unwrap();
```

File: docs/src/testing/contracts.md [hunk 26]
```diff
@@ -147,7 +146,7 @@ Using `SafeDispatcher` we can test that the function in fact panics with an expe
 fn handling_errors() {
     // ...
     
-    let contract_address = deploy(prepared).unwrap();
+    let contract_address = contract.deploy(calldata).unwrap();
```
Comment (piotmag769) [inline docs/src/testing/contracts.md:149]:
```suggestion
    let contract_address = contract.deploy(@calldata).unwrap();
```

File: docs/src/appendix/forge-library/deploy.md [hunk 27]
```diff
@@ -1,31 +1,34 @@
...
   - `class_hash` - class hash of a previously [declared](declare.md) contract
-  - `constructor_calldata` - snapshot of calldata for the constructor
+
+- `constructor_calldata` - snapshot of calldata for the constructor
 
 ```rust
 use array::ArrayTrait;
 use result::ResultTrait;
-use cheatcodes::PreparedContract;
+use cheatcodes::{ declare, ContractClassTrait };
```
Comment (piotmag769) [inline docs/src/appendix/forge-library/deploy.md:20]: Does it work without importing `ContractClassTraitImpl`?

Comment (piotmag769) [reply docs/src/appendix/forge-library/deploy.md:20]: Nvm, it seems like it does in similar case for me

File: cheatcodes/src/cheatcodes.cairo (hunk 18 above)
Comment (tkumor3) [reply cheatcodes/src/cheatcodes.cairo:51]:
```
fn _prepare_calldata(class_hash: @ClassHash, constructor_calldata: @Array::<felt252>) -> Array::<felt252>  {
    let class_hash: felt252 = (class_hash.clone()).into();
```
but I have to clone class_hash. Is it ok? @Arcticae

File: crates/cheatable-starknet/src/lib.rs (hunk 20 above)
Comment (tkumor3) [reply crates/cheatable-starknet/src/lib.rs:34]: the thing is, that we need to do it separately. Because when we want to precalculate contract_address we won't update `deploy_salt_base`. Maybe I will rename `generate_salt` to `get_salt` 🤔 what do you think? @Arcticae

File: crates/cheatable-starknet/src/lib.rs (hunk 3 above)
Comment (tkumor3) [reply crates/cheatable-starknet/src/lib.rs:35]: good idea-> it's less confusing!

File: docs/src/appendix/forge-library/deploy.md (hunk 27 above)
Comment (tkumor3) [reply docs/src/appendix/forge-library/deploy.md:20]: Yep @Arcticae suggested removing it from exports and it works

**Review by piotmag769 — COMMENTED** [review]
LGTM in general

File: crates/cheatable-starknet/src/cheatcodes/precalculate_address.rs [hunk 28]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+fn create_execute_calldata(calldata: &[Felt252]) -> Calldata {
+    let mut execute_calldata = vec![];
+    let mut calldata: Vec<StarkFelt> = calldata
+        .iter()
+        .map(|data| felt_to_stark_felt(data))
+        .collect();
+    execute_calldata.append(&mut calldata);
+    Calldata(execute_calldata.into())
```
[the remark this answers is not in the document]
Comment (tkumor3) [reply crates/cheatable-starknet/src/cheatcodes/precalculate_address.rs:51]: @piotmag769 but for deploy it's different function
```
fn create_execute_calldata(
    calldata: &[Felt252],
    class_hash: &ClassHash,
    account_address: &ContractAddress,
    entry_point_selector: &EntryPointSelector,
    salt: &ContractAddressSalt,
) -> Calldata 
```

File: crates/forge/src/cheatcodes_hint_processor.rs [hunk 29]
```diff
@@ -286,6 +283,23 @@ impl CairoHintProcessor<'_> {
                 print(inputs);
                 Ok(())
             }
+            "precalculate_address" => {
```
Comment (tkumor3) [inline crates/forge/src/cheatcodes_hint_processor.rs:286]: @war-in  I refactored `precalculate_address` to be coherent with your changes

File: docs/src/appendix/forge-library/precalculate_address.md [hunk 30]
```diff
@@ -0,0 +1,32 @@
...
+trait ContractClassTrait {
+    fn precalculate_address(self: @ContractClass, constructor_calldata: @Array::<felt252>) -> ContractAddress;
+    // ...
+}
+```
+Calculate the address of a contract in advance that will be returned upon the next deploy.
+
+- `self` - an instance of the struct `ContractClass` which is obtained by calling [declare](./declare.md).
+
+- `constructor_calldata` - snapshot of calldata for the deploy constructor. The constructor_calldata has an impact on the resulting contract_address. To precalculate the address for a future deploy, the calldata couldn't change.
```
Comment (piotmag769) [inline docs/src/appendix/forge-library/precalculate_address.md:13]:
```suggestion
- `constructor_calldata` - snapshot of calldata for the deploy constructor. The `constructor_calldata` has an impact on the resulting contract address. To precalculate an address for a future deployment, the calldata cannot change.
```

File: docs/src/appendix/forge-library/precalculate_address.md [hunk 31]
```diff
@@ -0,0 +1,32 @@
+# `precalculate_address`
+
+```rust
+trait ContractClassTrait {
+    fn precalculate_address(self: @ContractClass, constructor_calldata: @Array::<felt252>) -> ContractAddress;
+    // ...
+}
+```
+Calculate the address of a contract in advance that will be returned upon the next deploy.
```
Comment (piotmag769) [inline docs/src/appendix/forge-library/precalculate_address.md:9]:
```suggestion
Calculate an address of a contract in advance that would be returned when calling [`deploy`](./deploy.md).
```

File: docs/src/appendix/forge-library/precalculate_address.md [hunk 32]
```diff
@@ -0,0 +1,32 @@
...
+
+- `self` - an instance of the struct `ContractClass` which is obtained by calling [declare](./declare.md).
```
Comment (piotmag769) [inline docs/src/appendix/forge-library/precalculate_address.md:11]:
```suggestion
- `self` - an instance of the struct `ContractClass` which is obtained by calling [`declare`](./declare.md).
```

File: docs/src/appendix/forge-library/precalculate_address.md
[the remark this answers is not in the document]
Comment (piotmag769) [reply docs/src/appendix/forge-library/precalculate_address.md:1]: We should also add an info block informing that only the precalculated address is only correct for the very next deploy call and it will be different for all deploy calls after the first one after calculating the address

[the remark this answers is not in the document]
Comment (tkumor3) [reply docs/src/appendix/forge-library/precalculate_address.md:1]: I added it as warning

File: docs/src/appendix/forge-library/deploy.md [hunk 33]
```diff
@@ -1,31 +1,35 @@
...
+    // ...
+}
+```
 
 Deploys a contract and returns its address.
 
-- `prepared_contract` - an instance of the struct `PreparedContract` that consists of the following fields:
-  - `class_hash` - class hash of a previously [declared](declare.md) contract
-  - `constructor_calldata` - snapshot of calldata for the constructor
+- `self` - an instance of `ContractClass` struct can be obtained by invoking [declare](./declare.md).
```
Comment (piotmag769) [inline docs/src/appendix/forge-library/deploy.md]:
```suggestion
- `self` - an instance of `ContractClass` struct can be obtained by invoking [`declare`](./declare.md).
```

File: docs/src/appendix/forge-library/declare.md [hunk 34]
```diff
@@ -1,8 +1,16 @@
...
+Function [deploy](./deploy.md) and [precalculate_address](./precalculate_address.md) can be called on this struct.
```
Comment (piotmag769) [inline docs/src/appendix/forge-library/declare.md]:
```suggestion
Functions [`deploy`](./deploy.md) and [`precalculate_address`](./precalculate_address.md) can be called on this struct.
```

File: docs/src/appendix/forge-library/precalculate_address.md [hunk 35]
```diff
@@ -0,0 +1,37 @@
+# `precalculate_address`
+
+> ⚠️ **Warning**
+>
+> Precalculated address is only correct for the very next deploy call. It will be different for all deploy calls after the first one after calculating the address.
```
Comment (piotmag769) [inline docs/src/appendix/forge-library/precalculate_address.md:5]:
```suggestion
> Precalculated address is only correct for the very next [`deploy`](./deploy.md) call. It will be different for all `deploy` calls after the first one after calculating the address.
```

**Review by piotmag769 — APPROVED** [review]
Great work! 🔥 Please refrain from merging before release though 🙏
